### PR TITLE
PrestashopCollection needs a shop restriction

### LINKED ISSUE #28864: Inefficient queries in PrestaShopCollection, incoherent display of category names in breadcrumbs.
### Prerequisites

- [X] I understand and accept the project's [code of conduct](https://github.com/PrestaShop/PrestaShop/blob/develop/CODE_OF_CONDUCT.md)
- [X] I have already [searched in existing issues](https://github.com/PrestaShop/PrestaShop/issues?q=is%3Aissue+label%3ABug) and found no previous report of this bug.

### Describe the bug and add screenshots

There is a bug in PrestaShopCollection is inefficient in a multishop context when a language file `*_lang` is involved, and there is one language entry per shop.

Verifiable example: 
If a category has not the same name in all shops, the name displayed in the breadcrumb is always the name of the category in the main shop.

The reason behind: the join generated selects all lines, and not the ones related to the current shop. 

e.g. The query generated for the category breadcrumbs in `ProductController::getBreadcrumbLinks()` (`$categoryDefault->getAllParents()`) returns too many lines:
```
SELECT * FROM `ps_category` a0 LEFT JOIN `ps_category_lang` `a1` ON (a0.`id_category` = a1.`id_category`) WHERE (a0.`nleft` < 2328) AND (a0.`nright` > 2337) AND (a1.`id_lang` = 1) ORDER BY a0.`nleft` asc
```
Given `nl` as category level depth, and `ns` the number of shops, the query returns `nl * ns` lines whereas it should return only `nl` lines, one for each category.

Note: this is related to #28841, which is not reproducible.

### Expected behavior

I expect 
1. the query to be efficient
2. `PrestaShopCollection` returning the objects pertaining to the current shop and not the main shop.

### Steps to reproduce

In a multishop environment:
1. In the BO, select a shop different from the main shop, `shop A`
2. Go to `Catalog - Categories`
3. Edit the name of a parent category that has a child category
4. In the FO, in `shop A`, navigate to a product being member of the child category
5. Notice that the name of the parent category displayed in the breadcrumb is the one of the main shop, and not the one of `shop A`

PS: We have a 26 shops in our multishop environment, so the performance is also a problem, but it would not be possible to observe it directly.

### PrestaShop version(s) where the bug happened

1.7.8.6 - 1.7.7.8

### PHP version(s) where the bug happened

7.3

### If your bug is related to a module, specify its name and its version

_No response_

### PR DESCRIPTION
**Pull request by lmeyer1**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | Fixes the join of `multilang_shop` tables in `PrestShopCollection`. Both a problem of localization and performance.
| Type?             | bug fix
| Category?         | CO
| BC breaks?        | no
| Deprecations?     | no
| Fixed ticket?     | Fixes #28864.
| Related PRs       | -
| How to test?      | Follow the indications in #28864
| Possible impacts? | There should not be any

### CODE REVIEW

File: classes/PrestaShopCollection.php [hunk 1]
```diff
@@ -345,6 +344,9 @@ public function getAll($display_query = false)
             if ($this->id_lang) {
                 $this->where(self::LANG_ALIAS . '.id_lang', '=', $this->id_lang);
             }
+            if (!empty($this->definition['multilang_shop'])) {
+                $this->sqlWhere($this->join_list[self::LANG_ALIAS]['alias'] . '.`id_shop` = ' . Context::getContext()->shop->id);
```
Comment (matthieu-rolland) [inline classes/PrestaShopCollection.php]:
```suggestion
                $this->sqlWhere($this->join_list[self::LANG_ALIAS]['alias'] . '.`id_shop` = ' . (int) Context::getContext()->shop->id);
```

**Review by matthieu-rolland — COMMENTED** [review]
looks fine to me 👍  let's wait for the QA validation of your issue for the bug part

**Review by Progi1984 — APPROVED** [review]

**Review by ghost — APPROVED** [review]

**Comment by ghost** [comment]
This PR should be documented for people who have already fixed in their module.

Like [here](https://github.com/Invertus/elasticsearch/issues/9)

**Review by florine2623 — APPROVED** [review]
Hello @lmeyer1 ,

Tested with multistore, multi-language, categories with sub-categories, checked category name (in breadcrumb, in sitemap, on the top menu, on category page).
The name is well updated, there's no more confusion if using multistore.

It is QA ✅

**Comment by lmeyer1** [comment]
@florine2623 Thank you

**Review by matthieu-rolland — APPROVED** [review]

**Comment by matthieu-rolland** [comment]
thank you @lmeyer1 and @florine2623 👍

**Pull request merged**